python/cudf/cudf/_lib/pylibcudf/copying.pyx

```diff
@@ -55,3 +83,187 @@ cpdef Table gather(
             )
         )
     return Table.from_libcudf(move(c_result))
+
+
+cpdef Table table_scatter(Table source, Column scatter_map, Table target_table):
+    cdef unique_ptr[table] c_result
+
+    with nogil:
+        c_result = move(
+            cpp_copying.scatter(
+                source.view(),
+                scatter_map.view(),
+                target_table.view(),
+            )
+        )
+
+    return Table.from_libcudf(move(c_result))
+
+
+# TODO: Could generalize list to sequence
+cpdef Table scalar_scatter(list source, Column scatter_map, Table target_table):
+    cdef vector[reference_wrapper[const scalar]] source_scalars
+    source_scalars = _make_scalar_vector(source)
+
+    cdef unique_ptr[table] c_result
+    with nogil:
+        c_result = move(
+            cpp_copying.scatter(
+                source_scalars,
+                scatter_map.view(),
+                target_table.view(),
+            )
+        )
+
+    return Table.from_libcudf(move(c_result))
+
+
+cpdef object column_empty_like(Column input):
+    cdef unique_ptr[column] c_column_result
+    with nogil:
+        c_column_result = move(
+            cpp_copying.empty_like(
+                (<Column> input).view(),
+            )
+        )
+    return Column.from_libcudf(move(c_column_result))
+
+
+cpdef object table_empty_like(Table input):
+    cdef unique_ptr[table] c_table_result
+    with nogil:
+        c_table_result = move(
+            cpp_copying.empty_like(
+                (<Table> input).view(),
+            )
+        )
+    return Table.from_libcudf(move(c_table_result))
+
+
+cpdef Column allocate_like(
+    Column input_column, mask_allocation_policy policy, size=None
```

Comment:
Hmm this raises a good question of in what ways we want to deviate from libcudf APIs. Keyword-only arguments could be nice, but should we have systematic rules on when we use them?

Reply:
My preference is that optional arguments should be keyword-only, which is perhaps an easy systematic rule.

My rationale for this is that multiple defaulted keyword-like (but also positional) arguments are easy to get wrong when using an API. For example, `allocate_like` here takes an overriding `size`, but what if it also took an overriding `dtype`. With `allocate_like(input, policy, size=None, dtype=None)` I have to remember which way round to specify things if I want to override both, with `allocate_like(input, policy, *, size=None, dtype=None)` I don't, and can't get it wrong.

Equally, the use of the keyword name makes the (often untyped) argument self-documenting. It's obvious at the call-site that we're passing a mask_allocation_policy in as the policy argument (since that's an enum that we have to get a value out of), but for `size` it could come from anywhere.

Reply:
I agree, and @shwina also agreed offline. Unfortunately, Cython doesn't support keyword-only arguments for cdef functions, and therefore also not for cpdef functions.